standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -1855,43 +1869,81 @@ private MTable getMTable(String catName, String db, String table) {
       for (String t : tbl_names) {
         lowered_tbl_names.add(normalizeIdentifier(t));
       }
+
       query = pm.newQuery(MTable.class);
       query.setFilter("database.name == db && database.catalogName == cat && tbl_names.contains(tableName)");
       query.declareParameters("java.lang.String db, java.lang.String cat, java.util.Collection tbl_names");
-      Collection mtables = (Collection) query.execute(db, catName, lowered_tbl_names);
-      if (mtables == null || mtables.isEmpty()) {
-        // Need to differentiate between an unmatched pattern and a non-existent database
-        dbExistsQuery = pm.newQuery(MDatabase.class, "name == db && catalogName == cat");
-        dbExistsQuery.declareParameters("java.lang.String db, java.lang.String cat");
-        dbExistsQuery.setUnique(true);
-        dbExistsQuery.setResult("name");
-        String dbNameIfExists = (String) dbExistsQuery.execute(db, catName);
-        if (org.apache.commons.lang3.StringUtils.isEmpty(dbNameIfExists)) {
-          throw new UnknownDBException("Could not find database " +
-              DatabaseName.getQualified(catName, db));
+
+      if (projectionSpec == null) {
+        mtables = (List<MTable>) query.execute(db, catName, lowered_tbl_names);
+      }
+      else if(projectionSpec.getFieldList() != null && projectionSpec.getFieldList().size() > 1) {
+        // fetch partially filled tables using result clause
+        query.setResult(Joiner.on(',').join(projectionSpec.getFieldList()));
```

Comment:
Can we add a validation method before actually executing the query? That way you don't hit the database query if the projectionSpec has invalid fields and throw an error early. The performance is better and code becomes cleaner since you avoid the logic at 1924-1915 if you validate the fields first.